enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,214 @@
+MicroShift cluster.
+
+1. The cluster user deploys a Service of type: LoadBalancer 
+2. The loadbalancer service controller watches the creation of the service. It
+will update the Service's `.status.loadBalancer` field, and help the CNIs to 
+plumbing the ingress traffic to the endpoint pods.
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+This new controller is part of the microshift binary. It shall be able to know
+what CNI is used by the cluster, and behave differently based on the CNI type.
+
+### Risks and Mitigations
+
+The pluggable CNI design of MicroShift is still under discussion, we need to
+make sure that CNI type can be aware in MicroShift binary.
+
+### Drawbacks
+
+It is not a CNI agnostic design. The loadbalancer Service controller will behave
+differently between OVN-Kubernetes and other CNIs.
+
+## Design Details
+
+The Service of loadblancer type is designed to exposes the Service externally
+using a cloud provider's load balancer. Normally, there is a loadbalancer
+Service controller which is responsible for:
+
+1. provision a loadbalancer for the Service.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+However, in the context of MicroShift, as we only get one node in the cluster,
+it doesn't make sense to use a real load balancer to forward the ingress
+traffic. So instead of provisioning the loadbalancer instance, the controller
+is only responsible for helping the CNIs to plumbing the ingress traffic.
+
+### Traffic Plumbing
+
+Since different CNIs get different implementation on how to support Kubernetes
+Services. We need to do the traffic plumbing differently for different CNI. The
+controller shall be aware of the CNI type from the configuration and behave
+accordingly.
+
+#### OVN-Kubernetes
+
+OVN-Kubernetes has already implemented the [loadbalancer ingress
+IP](https://github.com/openshift/ovn-kubernetes/blob/master/docs/external-ip-and-loadbalancer-ingress.md)
+support. The ovnkube-node component which is running on each host watches the
+Service's `.status.loadBalancer` field, and insert iptables rules that does DNAT
+to the clusterIP of the Service. Therefore, from the loadbalancer Service
+controller perspective, it will only be responsible for the updating the status
+of the Service. The ovn-kubernetes will do the rest of the job.
+
+#### Other CNIs
+
+For other CNIs, which don't have the similar function as OVN-Kubernetes, neither
+do they bypass the kernel for the ingress traffic, we can take the similar
+approach as K3S. The K3S's loadbalancer Service controller provisions a
+[klipper-lb](https://github.com/k3s-io/klipper-lb) pod to each nodes for each
+loadbalancer Service. The pod can insert iptables rules which DNAT ingress
+traffic to the nodeIP of the Service.
+
+Another option is manipulating the iptables rule by this controller itself.
+Since there is only one node, and the microshsift binary has the access to the
```

Comment:
We don't plan multi-node right now but we don't want to close that path.
In case of multi-node this would be more difficult to maintain, right? Synchronize them across nodes (maybe with a CRD), but still - klipper-lb's approach seems easier

Reply:
Do you know how `kube-proxy` compares to `klipper-lb` resource wise? You wrote in other comment that `kube-proxy` is heavier, but how much? MicroShift is intended for edge deployments so I guess every MB of memory is worth gold

Reply:
As klipper-lb just runs once and then sleeps. Resource consumption can be seen as near zero. kube-proxy consumes ~30MB of memory in my kind cluster. However, we shall also keep in mind that some CNI doesn't support normal k8s service without kube-proxy. We'll have to run kube-proxy for those CNIs.

Reply:
I don't think we need to resolve this question immediately. We have several options, and we can consider them when we fully understand the requirements based on which other CNI implementations are are asked to support. I've retitled this section to indicate that it refers to future work.